Shell check

vreuter opened the pull request:
Just the encapsulation aspect of https://github.com/epigen/pypiper/pull/28 plus completion of `PipelineManager` docstring

nsheff commented:
looks fine to me, did you run the unit tests on it?

vreuter commented:
Yeah, they're broken. But they're also broken on `dev`. On `master` they're super slow, or at least the last couple assertions are; is that normal? I waited a few minutes but then killed it.

nsheff commented:
tests work for me fine on `dev`... `python test_pypiper.py`. On master it takes 5.8 seconds to complete successfully. but this branch does not pass tests, they fail.

nsheff commented on pypiper/pypiper.py:
```diff
@@ -480,35 +498,30 @@ def checkprint(self, cmd, shell="guess", nofail=False):
 				print("ERROR: Subprocess returned nonzero result, but pipeline is continuing because nofail=True")
+			# TODO: return nonzero, or something...
 
-		return returnvalue
```
I believe this is causing the test failure in this branch.

nsheff commented:
I can't figure out why it's failing the tests, but here's the error:
```
Traceback (most recent call last):
  File "test_pypiper.py", line 59, in test_me
    self.pp.run(cmd, lock_name="sleep")
  File "/home/sheffien/code/pypiper/pypiper/pypiper.py", line 456, in run
    process_return_code, local_maxmem = self.callprint(cmd, shell, nofail, container)  # Run command
  File "/home/sheffien/code/pypiper/pypiper/pypiper.py", line 627, in callprint
    self.fail_pipeline(e)
  File "/home/sheffien/code/pypiper/pypiper/pypiper.py", line 969, in fail_pipeline
    raise e
OSError: [Errno 2] No such file or directory

```

Something is messing up in the call to `pm.run` the command `echo hello` -- and if I change the settings for `shell` it still fails.

nsheff commented:
ah, got it. your conditional is breaking the splitting of commands:

```
		if not shell and likely_shell:
			print("Should this command run in a shell instead of directly in a subprocess?")
			#cmd = cmd.split()
			cmd = shlex.split(cmd)
```

+1 for unit tests! You are not splitting a command that is not shell and should not be shell. I'm also a bit lost on the meaning/naming of  "likely_shell", can you make that a bit more intuitive?

vreuter commented:
I named it based on the `print` statement that appears to be suggesting use of `shell`; I don't know why that's being suggested but am happy to rename. What do you think?

vreuter commented:
Why are the ones on `dev` broken? It's not just this branch.

nsheff commented on pypiper/pypiper.py:
```diff
@@ -461,6 +472,12 @@ def run(self, cmd, target=None, lock_name=None, shell="guess", nofail=False, cle
 
+
+	@staticmethod
+	def _likely_shell(cmd):
```
maybe let's call this somethign a bit more descriptive; maybe `_is_likely_shell()` ? or better, `check_shell()` or `shell_check_command()`, since it's not a method on a "Command" object but a static, which is what `likely_shell` seems to fit better with.

vreuter replied:
OK cool I like `check_shell` too; that's what I've changed it to.

nsheff commented:
`dev` tests work. I don't know, are you running them differently from me? They work fine for me on a brand new clone.

> tests work for me fine on dev... python test_pypiper.py. On master it takes 5.8 seconds to complete successfully. but this branch does not pass tests, they fail.

nsheff commented:
So, to be totally clear: Unit tests are working fine on both `dev` and `master`, but failing here because the logic for determining a shell command is now incorrect; non-shell commands like `echo hello` are not correctly being split into `['echo', 'hello']` and so `Call` is returning an error, as expected. Your PR is broke the logic on assessing shell commands as you attempted to abstract away the shell check. I'm sure it can be fixed...

vreuter commented:
Yeah you're right apparently I had stale dev on Rivanna; mine haven't completed, though, going on > 10 mins now; how fast do they run for you?

nsheff commented:
5.8 seconds...
but if they failed previously, you may have to delete the output folder, which is stale on a failed attempt (sorry)

vreuter commented:
Ah 👌 thanks will try that & see then...sorry for the confusion. Tsk tsk on me for having stale branch 😮

vreuter commented:
Hmm mine still fails actually.
```
##### [Epilogue:]
*   Total elapsed time:  0:00:00
*     Peak memory used:  0.0 GB
* Pipeline completed at:  (03-25 23:01:10) elapsed:0:00:00 _TIME_
Removing pipeline_output/
E
======================================================================
ERROR: test_me (__main__.PypiperTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_pypiper.py", line 28, in tearDown
    shutil.rmtree(self.pp.pipeline_outfolder)
  File "/share/apps/python/2.7.6/lib/python2.7/shutil.py", line 252, in rmtree
    onerror(os.remove, fullname, sys.exc_info())
  File "/share/apps/python/2.7.6/lib/python2.7/shutil.py", line 250, in rmtree
    os.remove(fullname)
OSError: [Errno 16] Device or resource busy: 'pipeline_output/.nfs000000003014c19b00000072'

----------------------------------------------------------------------
Ran 1 test in 6.369s

FAILED (errors=1)

Pypiper terminating spawned child process 117217
```

vreuter commented:
Is faster now, though. Might've just been ephemeral interactive node lag

vreuter commented:
Even stranger, that was for inside virtualenv on Rivanna. Different failure for outside venv locally:
```
##### [Epilogue:]
*   Total elapsed time:  0:00:04
*     Peak memory used:  0.0 GB
* Pipeline completed at:  (03-25 23:06:09) elapsed:0:00:04 _TIME_
E
======================================================================
ERROR: test_me (__main__.PypiperTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_pypiper.py", line 63, in test_me
    self.pp._wait_for_process(self.pp.running_subprocess)
  File "/Users/Vince/code/pypiper/pypiper/pypiper.py", line 630, in _wait_for_process
    local_maxmem = max(local_maxmem, self._memory_usage(p.pid) / 1e6)
  File "/Users/Vince/code/pypiper/pypiper/pypiper.py", line 1173, in _memory_usage
    status = open(proc_spot)
IOError: [Errno 2] No such file or directory: '/proc/47443/status'

======================================================================
ERROR: test_me (__main__.PypiperTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_pypiper.py", line 26, in tearDown
    self.pp3.stop_pipeline()
AttributeError: 'PypiperTest' object has no attribute 'pp3'

----------------------------------------------------------------------
Ran 1 test in 5.723s

FAILED (errors=2)

Pypiper terminating spawned child process 47401
```

vreuter commented:
Maybe an installation thing? Did you install after you cloned? Wouldn't matter I guess for the test script itself, but maybe it's importing module(s)/class(es) from an old install?

vreuter commented:
OK it's solved now, at least within the Travis build; the `dev` on works there, too. I don't know what's up with the interactive node I'm using on Rivanna, as both `dev` and  this branch fail on Rivanna but work for the Travis build. The failure for me locally was OSX-related I think. I updated the function name and fixed the change in the logic that had induced the failure.
